### PR TITLE
[GHA] Speed up PyTorch Layer unit tests

### PR DESCRIPTION
**Pull request by mryzhov**
### Details:
 - Enabled parallel call of PyTorch Layer tests on Linux
 - Fixed tests to be compatible with pytest-xdist plugin (moved random tensor generation inside the tests )

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Review by eaidova — APPROVED** [review]

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py [hunk 1]
```diff
@@ -34,17 +33,17 @@ def forward(self, input_tensor):
     @pytest.mark.precommit
     @pytest.mark.precommit_ts_backend
     @pytest.mark.precommit_fx_backend
-    def test_adaptive_avg_pool3d(self, ie_device, precision, ir_version, input_tensor, output_size):
-        self.input_tensor = input_tensor
+    def test_adaptive_avg_pool3d(self, ie_device, precision, ir_version, input_shape, output_size):
+        self.input_tensor = np.random.random_sample(input_shape).astype(np.float32)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py:37]: Why `random_sample`?

**Review by mvafin — CHANGES_REQUESTED** [review]
Changing from `np.random.randn` to `np.random.random_sample` is not good. Previously negative numbers were also created, while `random_sample` create numbers in range [0,1)

File: tests/layer_tests/pytorch_tests/test_all.py [hunk 2]
```diff
@@ -34,32 +34,35 @@ class TestAll(PytorchLayerTest):
...
-        np.random.randint(0, 2, (5, 9, 7)),
-        np.random.randint(0, 2, (10, 13, 11)),
-        np.random.randint(0, 2, (8, 7, 6, 5, 4)),
-        np.random.randint(0, 2, (11, 11), dtype=np.uint8),
-        np.random.randint(0, 2, (7, 7), dtype=np.uint8),
+    @pytest.mark.parametrize("input_shape, d_type", [
+        (np.eye(5,5), np.int16),
+        (np.zeros((5, 5)), np.int16),
+        (np.zeros((9,8)) + 1, np.int16),
+        ([5, 9, 7], np.int16),
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_all.py:41]: Why `int16`? Previously it as `int64` which is also not great, better would be `int32`

Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_all.py:41]: yes, missed that, should be int64 (as it was by default)

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py (hunk 1 above)
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py:37]: it accepts list as argument

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -907,6 +907,9 @@ jobs:
             python3 -m pip install $ov_dev_wheel_name[mxnet,caffe,kaldi,onnx,tensorflow2,pytorch]
           popd
 
+          # To enable pytest parallel features
+          python3 -m pip install pytest-xdist[psutil]
```
Comment (mvafin) [inline .github/workflows/linux.yml:911]: Shouldn't this be added to `requirements.txt` of tests? https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/requirements.txt

Comment (mryzhov) [reply .github/workflows/linux.yml:911]: I think no, it should be a part of CI, because user could run those tests without this plugin.

File: tests/layer_tests/pytorch_tests/test_all.py (hunk 2 above)
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_all.py]: fixed

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py (hunk 1 above)
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py]: changed with set seed argument, jfyi, it increased the execution time on 1minute

File: .github/workflows/linux.yml (hunk 3 above)
Comment (mryzhov) [reply .github/workflows/linux.yml:923]: To add or not add additional dependency we should answer what will not work without it?
imho this is user's specific tool and he should decide to install it or not, depends on his decision use it or not

btw, why we do not include CI system dependencies (git, ca-certificates) to install_dependencies.sh script then?

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py (hunk 1 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py]: You create generator in each test case, which make sense for parallel, but can be the reason for this degradation. And tbh I was thinking we will create just one generator with seed for all test suite. Also it complicates each test, so this actually change my mind, would it be acceptable to revert using generators with seed back?

Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py]: reverted

File: tests/layer_tests/pytorch_tests/test_all.py [hunk 4]
```diff
@@ -72,8 +75,13 @@ def test_all_noparams(self, input_tensor, ie_device, precision, ir_version):
     @pytest.mark.precommit
     @pytest.mark.xfail(condition=platform.system() == 'Darwin' and platform.machine() == 'arm64',
                        reason='Ticket - 122715')
-    def test_all(self, input_tensor, keepdim, ie_device, precision, ir_version):
-        self.input_tensor = input_tensor
-        for dim in range(len(input_tensor.shape)):
-            self._test(aten_all(dim, keepdim), None, "aten::all", 
+    def test_all(self, input_shape, d_type, keepdim, ie_device, precision, ir_version):
+        if type(input_shape) is list:
+            self.input_tensor = np.random.randint(0, 2, input_shape, dtype=d_type)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_all.py:80]: Can we avoid adding such ifs?

File: tests/layer_tests/pytorch_tests/test_repeat_interleave.py [hunk 5]
```diff
@@ -12,9 +12,8 @@
                                         {'repeats': 2, 'dim': 2},
                                         {'repeats': [2, 3], 'dim': 1},
                                         {'repeats': [3, 2, 1], 'dim': 3},
-                                        {'repeats': [3, 2, 1], 'dim': 3},
                                         {'repeats': 2, 'dim': None},
-                                        {'repeats': [random.randint(1, 5) for _ in range(36)], 'dim': None}))
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_repeat_interleave.py:17]: yes, and later we generate this list:
        if type(repeats) is list and len(repeats) == 1:
            repeats = np.random.randint(low=1, high=6, size=repeats[0], dtype='uint8')

File: tests/layer_tests/pytorch_tests/test_index_put_.py [hunk 6]
```diff
@@ -135,39 +135,42 @@ def forward(self, input_tensor, values, indices_0, indices_1):
...
     @pytest.mark.parametrize(
         "indices",
         (
-            (np.random.randint(low=0, high=2, size=(1,)), np.random.randint(low=0, high=2, size=(1,))),
-            (np.random.randint(low=0, high=2, size=(2,)), np.random.randint(low=0, high=2, size=(2,))),
-            (np.array([0, 1, 0]), np.array([1, 1, 0])),
-            (np.ones(shape=(3,)), np.ones(shape=(3,))),
-            (np.ones(shape=(3,)), np.zeros(shape=(3,))),
+            [[1, ], [1, ]],
+            [[2, ], [2, ]],
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_index_put_.py:151]: do you suggest to remove those tensors?

File: tests/layer_tests/pytorch_tests/test_all.py (hunk 4 above)
Comment (mryzhov) [reply tests/layer_tests/pytorch_tests/test_all.py:80]: we need to decide inside of test should we generate tensor data or it is already generated and we should use it as is.
 we can change the test arguments (for example, we can use dict), but it will change if condition only

File: tests/layer_tests/pytorch_tests/test_repeat_interleave.py [hunk 7]
```diff
@@ -39,6 +38,9 @@ def forward(self, input_tensor):
     @pytest.mark.precommit
     def test_repeat_interleave_const_repeats(self, ie_device, precision, ir_version, input_data):
         repeats = input_data['repeats']
+        if type(repeats) is list and len(repeats) == 1:
+            repeats = np.random.randint(low=1, high=6, size=repeats[0], dtype='uint8')
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_repeat_interleave.py:42]:
```suggestion
            repeats = random.randint(1, 5) for _ in range(repeats[0])
```

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py [hunk 8]
```diff
@@ -23,8 +22,8 @@ def __init__(self, output_size) -> None:
                 super().__init__()
                 self.output_size = output_size
 
-            def forward(self, input_tensor):
-                return torch.nn.functional.adaptive_avg_pool3d(input_tensor, self.output_size)
+            def forward(self, input_shape):
+                return torch.nn.functional.adaptive_avg_pool3d(input_shape, self.output_size)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py:26]:
```suggestion
            def forward(self, input_tensor):
                return torch.nn.functional.adaptive_avg_pool3d(input_tensor, self.output_size)
```

File: tests/layer_tests/pytorch_tests/test_all.py [hunk 9]
```diff
@@ -72,8 +75,13 @@ def test_all_noparams(self, input_tensor, ie_device, precision, ir_version):
...
+        else:
+            self.input_tensor = input_shape
+        self._test(aten_all_noparam(), None, "aten::all",
+                ie_device, precision, ir_version, trace_model=True, freeze_model=False)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_all.py:84]:
```suggestion
```

**Review by mvafin — APPROVED** [review]
Small comments

File: tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py [hunk 10]
```diff
@@ -8,13 +8,12 @@
 from pytorch_layer_test_class import PytorchLayerTest
 
 
-@pytest.mark.parametrize('input_tensor', (np.random.randn(1, 2, 8, 9, 10).astype(np.float32),
-                                          np.random.randn(2, 8, 9, 10).astype(np.float32)))
-@pytest.mark.parametrize('output_size', ([5, 7, 9], 7))
+@pytest.mark.parametrize('input_shape', [[1, 2, 8, 9, 10], [2, 8, 9, 10]])
```
Comment (mryzhov) [inline tests/layer_tests/pytorch_tests/test_adaptive_avg_pool.py:11]:
```suggestion
@pytest.mark.parametrize('input_tensor', [[1, 2, 8, 9, 10], [2, 8, 9, 10]])
```

**Review by ilya-lavrenov — APPROVED** [review]
